gerrit_python_tools/gerrit.py

```diff
@@ -624,13 +682,30 @@ def send_upstream(self, downstream, upstream):
         approvals = self.get_approvals(downstream.SSH())
 
         # Check to see if comment has necessary approvals.
-        if not self.is_upstream_approved(approvals):
-            msg = ("Could not send to upstream: One or more labels"
-                   " not approved.")
+        if self.is_forced():
+            if not self.is_release_approved(approvals):
+                msg = "Could not send to upstream: Release not approved."
+                logger.debug("Change %s: %s" % (self.change_id, msg))
+                ssh.exec_once('gerrit review -m %s %s'
+                              % (pipes.quote(msg), self.revision))
+                return
+            msg = ("***WARNING***\n%s (%s) has requested a forced upstream "
+                   "push. Bypassing all votes except for Release...\n\n"
```

Comment:
I don't want to implement it. But it's a necessary evil. Other than the check that the release check is present there is no restriction on this use. Think of it as the ink packs that shoot on someone's hand when they pull the fire alarm. In the right situation it's fine and you did nothing wrong, but knowing that your hand will be stained and you will be easy to identify will act a deterrent to using it unless absolutely necessary.

Putting the warning after the invocation is also the only way this works since the system is asynchronous and there is no mechanism for "are you sure."  I want it to be clear in no uncertain terms that using this is not to be done lightly and should be used as rarely as possible. I also want it to be obvious when reading the comments that this mechanism was used.


----
Andrew Kerr
QA, OpenStack
Data Fabric Group

NetApp
919.476.4899 Direct Phone
Andrew.Kerr@netapp.com<mailto:Andrew.Kerr@netapp.com>

On Nov 18, 2015, at 12:21 PM, Goutham Pacha Ravi <notifications@github.com<mailto:notifications@github.com>> wrote:


In gerrit_python_tools/gerrit.py<https://github.com/NetApp-openstack-dev/gerrit-python-tools/pull/3#discussion_r45229857>:

> @@ -624,13 +682,30 @@ def send_upstream(self, downstream, upstream):
>          approvals = self.get_approvals(downstream.SSH())
>
>          # Check to see if comment has necessary approvals.
> -        if not self.is<http://self.is>_upstream_approved(approvals):
> -            msg = ("Could not send to upstream: One or more labels"
> -                   " not approved.")
> +        if self.is<http://self.is>_forced():
> +            if not self.is<http://self.is>_release_approved(approvals):
> +                msg = "Could not send to upstream: Release not approved."
> +                logger.debug("Change %s: %s" % (self.change_id, msg))
> +                ssh.exec_once('gerrit review -m %s %s'
> +                              % (pipes.quote(msg), self.revision))
> +                return
> +            msg = ("***WARNING***\n%s (%s) has requested a forced upstream "
> +                   "push. Bypassing all votes except for Release...\n\n"


akerr: Fair enough.. Audit Trail is an understandable need, blame pinning is not.. If you don't want people to use it, why implement it? If you are implementing it, the reason is people'll need to use it. Trust me, I would not 'mistakenly' type "force-send-upstream" if I didn't know what I was doing. And If I know what I'm doing, the last thing I want is this offending mistrust. [:-1:]

-
Reply to this email directly or view it on GitHub<https://github.com/NetApp-openstack-dev/gerrit-python-tools/pull/3/files#r45229857>.